### PR TITLE
mgmt: mcumgr: Add ability for MCUmgr to setup at boot-up

### LINKED ISSUE #52154: mgmt: mcumgr: Add Kconfig to automatically register handlers and mcumgr functionality
**Is your enhancement proposal related to a problem? Please describe.**
At present, mcumgr requires command handlers register their functionality manually and then mcumgr is started manually. This can be started using a sys init at bootup to simplify this

**Describe alternatives you've considered**
Leave registration as manual

### PR DESCRIPTION
**Pull request by nordicjm**
Adds an optional Kconfig that allows MCUmgr to set itself up at
    boot-up, register command handlers and initiate transports,
    instead of needing user-application interaction to perform this.

Fixes #52154

**Note** First commit is from https://github.com/zephyrproject-rtos/zephyr/pull/52356 and can be ignored

### CODE REVIEW

File: subsys/mgmt/mcumgr/transport/smp_udp.c [hunk 1]
```diff
@@ -279,3 +290,31 @@ int smp_udp_close(void)
...
+				      struct net_if *iface)
+{
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
+		LOG_INF("Network connected");
+
+		if (smp_udp_open() < 0) {
+			LOG_ERR("Could not open SMP UDP");
+		}
+
+		return;
```
Comment (aunsbjerg) [inline subsys/mgmt/mcumgr/transport/smp_udp.c:305]: Both the explicit return statements here are redundant and can be removed.

File: subsys/mgmt/mcumgr/transport/smp_udp.c [hunk 2]
```diff
@@ -279,3 +290,31 @@ int smp_udp_close(void)
...
+	} else if (mgmt_event == NET_EVENT_L4_DISCONNECTED) {
+		LOG_INF("Network disconnected");
+		smp_udp_close();
+		return;
+	}
+}
+
+void smp_udp_boot_init(void)
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply subsys/mgmt/mcumgr/transport/smp_udp.c]: Fixed

File: include/zephyr/mgmt/mcumgr/smp_udp.h [hunk 3]
```diff
@@ -17,19 +17,24 @@ extern "C" {
...
  *
- * @return 0 on success; negative error code on failure.
+ * @return	0 on success; negative error code on failure.
  */
 int smp_udp_close(void);
 
+/**
+ * @brief       Starts UDP transport at start-up if ``CONFIG_MCUMGR_SMP_BOOT_SETUP`` is enabled.
+ */
+void start_smp_bt(void);
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply include/zephyr/mgmt/mcumgr/smp_udp.h]: Fixed

**Comment by de-nordic** [comment]
The Zephyr group already does automatic registration, but uses different approach:
https://github.com/zephyrproject-rtos/zephyr/blob/b49b11555c9ae33a22e3cc0586a0ae31842507f2/subsys/mgmt/mcumgr/lib/cmd/zephyr_basic/src/basic_mgmt.c#L61-L70

**Comment by nordicjm** [comment]
> The Zephyr group already does automatic registration, but uses different approach:

Question is, should this be centralised so that all commands are registered at boot (and have a Kconfig for the transports to set them up automatically) using a single SYS_INIT() rather than multiple, or go with multiple?

**Comment by nordicjm** [comment]
Updated to move the zephyr basic management setup to the new system

**Pull request closed without merging**